SwiftSyntaxCLI/Sources/swift-syntax-cli/common/Builder.swift

```diff
@@ -0,0 +1,120 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+protocol Builder {
+  var arguments: BuilderArguments { get }
+}
+
+extension Builder {
+  func buildProduct(productName: String) throws {
+    print("** Building product " + productName + " **")
+    try build(packageDir: Paths.packageDir, productName: productName)
+  }
+
+  func buildExample(exampleName: String) throws {
+    print("** Building example " + exampleName + " **")
+    try build(packageDir: Paths.examplesDir, productName: exampleName)
+  }
+
+  private func build(packageDir: URL, productName: String) throws {
+    let args = ["--product", productName]
+
+    var env = ProcessInfo.processInfo.environment
+    env["SWIFT_BUILD_SCRIPT_ENVIRONMENT"] = "1"
+
+    if arguments.enableRawsyntaxValidation {
+      env["SWIFTSYNTAX_ENABLE_RAWSYNTAX_VALIDATION"] = "1"
+    }
+
+    if arguments.enableTestFuzzing {
+      env["SWIFTPARSER_ENABLE_ALTERNATE_TOKEN_INTROSPECTION"] = "1"
+    }
+
+    // Tell other projects in the unified build to use local dependencies
+    env["SWIFTCI_USE_LOCAL_DEPS"] = "1"
+    env["SWIFT_SYNTAX_PARSER_LIB_SEARCH_PATH"] = arguments.toolchain + "/lib/swift/macos"
+
+    try swiftpmInvocation(
+      action: "build",
+      packageDir: packageDir,
+      additionalArguments: args,
+      env: env
+    )
+  }
+
+  @discardableResult
+  func swiftpmInvocation(
+    action: String,
+    packageDir: URL,
+    additionalArguments: [String],
+    env: [String: String]
+  ) throws -> String {
+    var args = [action]
+    args += ["--package-path", packageDir.relativePath]
+    args += ["--scratch-path", arguments.buildDir]
+
+    #if !canImport(Darwin)
+    args += ["--enable-test-discovery"]
+    #endif
+
+    if arguments.release {
+      args += ["--configuration", "release"]
+    }
+
+    if let multirootDataFile = arguments.multirootDataFile {
+      args += ["--multiroot-data-file", multirootDataFile]
+    }
+
+    if arguments.disableSandbox {
+      args += ["--disable-sandbox"]
+    }
+
+    if arguments.verbose {
+      args += ["--verbose"]
+    }
+
+    args += additionalArguments
+
+    let process = Process()
+    let outputPipe = Pipe()
+    process.environment = env
+    process.executableURL = URL(fileURLWithPath: arguments.toolchain).appendingPathComponent("bin").appendingPathComponent("swift")
+    process.arguments = args
+    process.standardOutput = outputPipe
+
+    let outputHandle = outputPipe.fileHandleForReading
+    outputHandle.waitForDataInBackgroundAndNotify()
```

Comment:
Why do you need to call `waitForDataInBackgroundAndNotify`? I’ve never seen this needed before (which isn’t to say that it’s not needed).